src/airy.jl

```diff
@@ -1,107 +1,523 @@
-    isnan(x) && return x
-    x_abs = abs(x)
-    z = 2 * x_abs^(T(3)/2) / 3
+    airy_large_argument_cutoff(z) && return airyai_large_argument(z)
+    airyai_power_series_cutoff(x, y) && return airyai_power_series(z)
 
     if x > zero(T)
-        return isinf(z) ? zero(T) : sqrt(x / 3) * besselk(one(T)/3, z) / π
-    elseif x < zero(T)
-        Jv, Yv = besseljy_positive_args(one(T)/3, z)
-        Jmv = (Jv - sqrt(T(3)) * Yv) / 2
-        return isinf(z) ? throw(DomainError(x, "airyai(x) is not defined.")) : sqrt(x_abs) * (Jmv + Jv) / 3
-    elseif iszero(x)
-        return T(0.3550280538878172)
+        # use relation to besselk (http://dlmf.nist.gov/9.6.E1)
+        zz = 2 * z^(T(3)/2) / 3
+        return sqrt(z / 3) * besselk_continued_fraction_shift(one(T)/3, zz) / π
+    else
+        # z is close to the negative real axis
+        # for imag(z) == 0 use reflection to compute in terms of bessel functions of first kind (http://dlmf.nist.gov/9.6.E5)
+        # use computation for real numbers then convert to input type for stability
+        # for imag(z) != 0 use rotation identity (http://dlmf.nist.gov/9.2.E14)
+        if iszero(y)
+            xabs = abs(x)
+            xx = 2 * xabs^(T(3)/2) / 3
+            Jv, Yv = besseljy_positive_args(one(T)/3, xx)
+            Jmv = (Jv - sqrt(T(3)) * Yv) / 2
+            return convert(eltype(z), sqrt(xabs) * (Jmv + Jv) / 3)
+        else
+            return exp(pi*im/3) * _airyai(-z*exp(pi*im/3))  + exp(-pi*im/3) * _airyai(-z*exp(-pi*im/3))
+        end
     end
 end
 
 """
     airyaiprime(x)
 Derivative of the Airy function of the first kind ``\\operatorname{Ai}'(x)``.
 """
-airyaiprime(x::Real) = _airyaiprime(float(x))
+airyaiprime(z::Number) = _airyaiprime(float(z))
 
 _airyaiprime(x::Float16) = Float16(_airyaiprime(Float32(x)))
+_airyaiprime(z::ComplexF16) = ComplexF16(_airyaiprime(ComplexF32(z)))
+
+function _airyaiprime(z::ComplexOrReal{T}) where T <: Union{Float32, Float64}
+    if ~isfinite(z)
+        isnan(z) && return z
+        if abs(angle(z)) < 2π/3
+            return -exp(-z)
+        else
+            return 1 / z
+        end
+    end
+    x, y = real(z), imag(z)
+    zabs = abs(z)
 
-function _airyaiprime(x::T) where T <: Union{Float32, Float64}
-    isnan(x) && return x
-    x_abs = abs(x)
-    z = 2 * x_abs^(T(3)/2) / 3
+    airy_large_argument_cutoff(z) && return airyaiprime_large_argument(z)
+    airyai_power_series_cutoff(x, y) && return airyaiprime_power_series(z)
 
     if x > zero(T)
-        return isinf(z) ? zero(T) : -x * besselk(T(2)/3, z) / (sqrt(T(3)) * π)
-    elseif x < zero(T)
-        Jv, Yv = besseljy_positive_args(T(2)/3, z)
-        Jmv = -(Jv + sqrt(T(3))*Yv) / 2
-        return isinf(z) ? throw(DomainError(x, "airyaiprime(x) is not defined.")) : x_abs * (Jv - Jmv) / 3
-    elseif iszero(x)
-        return T(-0.2588194037928068)
+        # use relation to besselk (http://dlmf.nist.gov/9.6.E2)
+        zz = 2 * z^(T(3)/2) / 3
```

Comment:
how does `z^(T(3)/2)` compare to `sqrt(z^3)`? I'd expect the latter to be faster (if slightly less accurate)

Reply:
`sqrt(z)^3`is also worth checking as it prevents overflow.

Reply:
Also worth considering is `z*sqrt(z)`

Reply:
O yes  - I like that much better. Will update. Hard to tell much about speed because this is within a pretty complex branch but preventing overflow is good here.